### ISSUE TITLE
NuxtHub  integration

### ISSUE THREAD

**Issue by husayt** [pruvious/pruvious#37]
I just stumbled on this project and what a nice surprise. Looks really promising. I will definitely give it a try.

One suggestion I have is [Nuxthub](https://admin.hub.nuxt.com/) integration. That will provide one click, seamless deployment and hosting. Currently one needs to setup database locally/externally to get started. With Nuxthub taking care of this it will remove big friction point.

From what I see it should be not a difficult task, as Nuxthub supports Sqlite (D1), blobs, cache, DrizzleOrm. 
That alone will make this such a powerful stack

@murisceman this project is an excellent and much needed initiative. Thank you for all the effort

**Comment by murisceman** [comment]
Thank you for your suggestion, @husayt. I have already reviewed the scope of work required to adapt Pruvious to function on CloudFlare Workers. Here's a summary:

- Replace Sequelize with Kysely or Drizzle (WIP: #14)
  > Needs to support D1.
- Figure out how to optimize images (https://github.com/nuxt-hub/core/issues/131)
  > Currently Pruvious uses [Sharp](https://github.com/topics/sharp-js), which doesn't work on the edge.
- Add bcrypt support (https://github.com/dcodeIO/bcrypt.js)
  > Besides [Argon2](https://github.com/ranisalt/node-argon2), which is the only password hashing algorithm in Pruvious (and does not work on CFW), a reasonable solution would be to add an option for bcrypt.

As simple as it sounds, a significant amount of work (~100-200 hours) is required to ensure that everything runs smoothly. Recently, I've had just enough time to fix bugs and add minor features to Pruvious, so this will have to wait until my main job becomes less demanding.

**Comment by husayt** [comment]
Absolutely. This might look easy on the surface, but there might be  a lot of parts needed to change to bring this to work. @murisceman thanks for mentioning some of them.

Might be good to find the minimal scope for this work, as investing 200 hours upfront might be too much time one might want to put into this at first. 

@murisceman would you be open to have a chat to discuss some ideas and suggestions I have. If you could drop me your email (my email in my profile) we could take it from there.


Thank you

**Comment by murisceman** [comment]
I would appreciate if you could share your ideas and suggestions here, so other users could join the conversation. Thank you!

**Comment by husayt** [comment]
@murisceman what I wanted to discuss was bit wider than this project. 

I became very intrigued by the story you provided in producthunt.  Your history with wordpress, envato, templates, issues you described about wordpress templates drew a lot of parallels to my experience. We have been working on something very related, but it is not yet open to public. That is why I wanted to discuss that with you separately.

Hopefully it might open some exciting directions

**Comment by murisceman** [comment]
Hey @husayt, you can contact me privately at support@pruvious.com. Cheers!

**Comment by MineDrum** [comment]
I see that this issue is closed and that there is much work to be done, but wanted to add that I would also love this support. 

I think the optimized images could be a lower priority in the beginning. 
Is there anything I can do to help?

**Comment by murisceman** [comment]
Thank you for offering your help, @MineDrum.

**Comment by husayt** [comment]
@murisceman this could be helpful I guess 

https://github.com/atinux/nuxt-bcrypt-edge

**Comment by Flexusma** [comment]
@murisceman I can take a look if I find some time, can't make any guarantees though. Also have a job and my own company to look after.

**Comment by murisceman** [comment]
Hey @Flexusma and @husayt, thank you both for the link and for offering to help with this. I have already checked out this package, and it works reasonably well. The performance isn't the best, but it's acceptable for now. I hope there will be a more optimized way to hash passwords on Cloudflare Workers in the future.

Here's a demo: https://bcrypt.nuxt.dev/

I'm considering setting a default of 12 hashing rounds, which takes about 500-600 ms on a Worker.

**Comment by vanling** [comment]
I'm an amateur when it comes to hashing, but what about https://github.com/unjs/uncrypto? 

_Found this issue/topic after I had some build pipeline issues. For Nuxt projects, when I push to Git, an automatic build triggers on an ARM-based Mac Mini. Only the .output folder is deployed (expected for Nuxt3). However, it seems argon2 built on macOS/ARM doesn't run on Linux/x86 or other platforms._

**Comment by murisceman** [comment]
Hey @vanling, PBKDF2 (Web Crypto) was my last resort for this. Fortunately, we can now use bcrypt, which is a much better option for password hashing.

https://github.com/unjs/uncrypto/issues/40

**Comment by murisceman** [comment]
Hey @LTDCoaching, here is an update:

-  ✅ ORM - created one from scratch that supports SQLite, PostgreSQL, and D1. It works like a charm in Cloudflare Workers.
-  ✅ Auth - solved using `bcrypt-edge`, and I have prepared the codebase for future hashing algorithms.
-  ❌ Images - this is still on my to-do list. Planning to explore pure JS image optimization solutions.

**Comment by crearis** [comment]
Wow Muris: This is already 2/3 of climbing a really high mountain!! Unbelievable how you are focussed on solving this task!!

**Comment by sambedingfield** [comment]
+1 for NuxtHub

@murisceman any plans on enabling GitHub Sponsorships (or any other means of donations) so we can help support/show our gratitude?

**Comment by murisceman** [comment]
Hey @sambedingfield, thanks for asking! I'll look into this after v4 launch 😊

**Comment by crearis** [comment]
Hey @sambedingfield, thanks this is very important!!
Meanwhile we stand on the sidelines, watch and as 'who is faster?' 
Muris on this climb with v4 or the community with adding another 500 stars ;)
Go on folks, star this!!!
